driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -997,4 +1075,63 @@ private long awaitNanos(final long timeoutNanos) {
         private MutableReference() {
         }
     }
+
+    private ServerStats getServerStats(final ObjectId serviceId) {
+        ServerStats serverStats = serverStatsMap.get(serviceId);
+        if (serverStats == null) {
+            serverStats = new ServerStats();
+            serverStatsMap.put(serviceId, serverStats);
+        }
+        return serverStats;
+    }
+
+    private synchronized void addConnectionToServerStats(final ObjectId serviceId) {
+        ServerStats serverStats = getServerStats(serviceId);
+        serverStats.incrementCount();
+    }
+
+    private synchronized void removeConnectionFromServerStats(final ObjectId serviceId) {
+        ServerStats serverStats = serverStatsMap.get(serviceId);
+        // assert not null
+        serverStats.decrementCount();
+        if (serverStats.getConnectionCount() == 0) {
+            serverStatsMap.remove(serviceId);
+        }
+    }
+
+    private synchronized void incrementGenerationInServerStats(final ObjectId serviceId) {
+        ServerStats serverStats = getServerStats(serviceId);
+        // assert not null
+        serverStats.incrementGeneration();
+    }
+
+    private synchronized int getGenerationFromServerStats(final ObjectId serviceId) {
+        ServerStats serverStats = getServerStats(serviceId);
+        return serverStats.getGeneration();
+    }
+
+    private static class ServerStats {
+        private int generation;
+        private int connectionCount;
```

Comment:
Include the connectionCount so that we know when we can remove the value from the map